src/stash/src/postgres.rs

```diff
@@ -716,53 +791,112 @@ impl Stash {
                 counts = format!("{:?}", counts.lock().unwrap()),
             );
         }
-        client
-            .batch_execute(tx_end)
-            .await
-            .map_err(|err| TransactionError::Commit(err.into()))?;
 
-        fail_point!("stash_commit", |r| Err(TransactionError::Commit(
-            r.unwrap_or_else(|| "stash_commit failpoint".to_string())
-                .into()
-        )));
+        let committed_if_version: i64 = epoch_row.get("version");
+
+        // We can't use the failpoint macro here because we need to move `res`
+        // into the error return, but because `res` is generic and a T, we can't
+        // create one. Calling `res.clone()` would require `T: Clone` which
+        // forces `Data` to have `Clone` which we maybe don't want (unclear).
+        // Thus, use the hidden function that the macro calls.
+
+        // Have both a pre and post commit failpoint to simulate each kind of
+        // error.
+        if let Some(_) = fail::eval("stash_commit_pre", |_| "") {
+            return Err(TransactionError::Commit {
+                committed_if_version,
+                result: res,
+            });
+        }
+
+        if let Err(_) = client.batch_execute(tx_end).await {
+            return Err(TransactionError::Commit {
+                committed_if_version,
+                result: res,
+            });
+        }
+
+        if let Some(_) = fail::eval("stash_commit_post", |_| "") {
+            return Err(TransactionError::Commit {
+                committed_if_version,
+                result: res,
+            });
+        }
 
         Ok(res)
     }
+
+    /// Reports whether a COMMIT that returned an error actually succeeded. An
+    /// Err return from this function is retryable normally (if
+    /// `!err.is_unrecoverable()`).
+    #[tracing::instrument(name = "stash::determine_commit", level = "debug", skip_all)]
+    async fn determine_commit(&mut self, committed_if_version: i64) -> Result<bool, StashError> {
```

Comment:
Do we need to worry about multiple concurrent stash transactions? If so would it be possible for a concurrent transaction to commit with our expected version and then we'd mistakenly think that our commit succeeded when it didn't?

Reply:
Nope, that shouldn't happen. A single stash has a single epoch and rust prevents it from mutably borrowing to start multiple txns like this.